### PR TITLE
Update to L1T Global for 2016 Menu (81X) - No Conditions Changes

### PR DESCRIPTION
**Pull request by mulhearn**
The history will be cleaned a bit (fewer but preserving developer commits) after it passes tests.  (I hate rebasing before I know the final product.)

This version removes the dependence on new CondFormats from L1TGlobalProducer while waiting on:  https://github.com/cms-sw/cmssw/pull/14201.

This brings the updated uGT software with bitwise correct emulation of correlation conditions.   Until cms-sw/cmsdist#2254 the correlation conditions will be empty, but it should otherwise run fine.

### CODE REVIEW

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
Ah dang, I forgot to include the HLT fix-up... this will die.

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
The addOn tests now all pass in my local test.  Probably this will pass all tests, unless I let in a regression...

File: DataFormats/L1TGlobal/BuildFile.xml [hunk 1]
```diff
@@ -5,3 +5,4 @@
 <export>
   <lib   name="1"/>
 </export>
+<flags CXXFLAGS="-g -D=EDM_ML_DEBUG"/>
```
Comment (fwyzard) [inline DataFormats/L1TGlobal/BuildFile.xml]: I guess we should remove this for the final integration ?

File: DataFormats/L1TGlobal/interface/GlobalAlgBlk.h [hunk 2]
```diff
@@ -92,12 +90,6 @@ class GlobalAlgBlk
 
 private:
 
-    /// orbit number
-    int m_orbitNr;
-
-    /// bunch cross number of the actual bx
-    int m_bxNr;
-
```
Comment (fwyzard) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: why were these removed from the uGT data format ?
what is an alternative method to monitor the orbit and bx in which the various L1 triggers are firing ?
do you know if the same information is available somewhere else, e.g. from the TCDS data ?

File: DataFormats/L1TGlobal/interface/GlobalAlgBlk.h [hunk 3]
```diff
@@ -109,9 +101,12 @@ class GlobalAlgBlk
     //Prescale Column
     int m_preScColumn;
 
+    //Menu UUID
+    int m_menuUUID;
+    int m_firmwareUUID;
```
Comment (fwyzard) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: I'm glad to see this addition. Do you happen to know if the same UUID should be available from the L1 menu in the EventSetup, to check for consistency ?

File: DataFormats/L1TGlobal/interface/GlobalAlgBlk.h [hunk 4]
```diff
@@ -109,9 +101,12 @@ class GlobalAlgBlk
...
    
     std::vector<bool> m_algoDecisionInitial;
-    std::vector<bool> m_algoDecisionPreScaled;
+    std::vector<bool> m_algoDecisionInterm;
```
Comment (fwyzard) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: out of curiosity, what is the purpose of this change ?
does it mean that the decision after prescales but before masking is no longer available ?

**Comment by fwyzard** [comment]
@mulhearn , could you clarify the impact of these changes ?
do we plan to make these changes online (i.e. need backport to 8.0.x) ?

File: DataFormats/L1TGlobal/BuildFile.xml (hunk 1 above)
Comment (mulhearn) [inline DataFormats/L1TGlobal/BuildFile.xml]: Ooops, thanks for catching that...

File: DataFormats/L1TGlobal/interface/GlobalAlgBlk.h (hunk 4 above)
Comment (mulhearn) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: It changes notation to reflect a hardware change.  We requested that initial decision be truly initial decision, the algorithm output with no mask / veto / pre-scale applied.  So now there is initial / final and one intermediate point which I can't recall the exact intermediate stage.

**Comment by mulhearn** [comment]
Yes, we need these in 80x to correctly emulate the uGT at the HLT.  I only don't have 80x PR first because #14141 isn't merged yet...

**Comment by mulhearn** [comment]
@blwiner, @tmatsush can you answer Andreas questions above more precisely?

File: DataFormats/L1TGlobal/interface/GlobalAlgBlk.h (hunk 3 above)
Comment (mulhearn) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: This is a feature of the new library, so I believe we will be able to compare the library as parsed by UTM with this unpacked value.  But Takashi et al can probably answer more precisely.

Comment (tmatsush) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: uGT data will contain 32-bit hash values generated from menu name and firmware-uuid
They will be generated with the following methods;
tmeventsetup::getMmHashN(L1TUtmTriggerMenu::getName())
tmeventsetup::getMmHashN(L1TUtmTriggerMenu:: getFirmwareUuid())

File: DataFormats/L1TGlobal/interface/GlobalAlgBlk.h (hunk 4 above)
Comment (tmatsush) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: Masks and prescales are applied in the following order;
  GTL = (1) => AlgoBxMask = (2) => Prescale = (3) => FinOR mask = (4) => 
At the moment uGT algorithms decisions are sampled at (2), (3) and (4)
The request was to change the sampling points to (1), (2) and (3)
In order to shift the sampling point, we need to stop using FinOR mask and
apply prescale value calculated with  prescale[ii] * finormask[ii]. (prescale value of 0 is equivalent to mask in uGT)
The firmeware and software need to be updated at the same time.

Comment (fwyzard) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: Understood, thanks !

File: DataFormats/L1TGlobal/interface/GlobalAlgBlk.h (hunk 3 above)
Comment (fwyzard) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: OK, thanks for the details.

This means we can add a check within the HLT and/or DQM that the menu used in the L1 and the menu from the EventSetup are the same.

File: DataFormats/L1TGlobal/interface/GlobalAlgBlk.h (hunk 2 above)
Comment (fwyzard) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: @tmatsush , is this a change coming to the hardware/firmware as well ?

Comment (tmatsush) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: uGT read-out block does not [0] and will not [1] contain orbit/bx number.
[0] https://twiki.cern.ch/twiki/bin/viewauth/CMS/GlobalTriggerUpgrade-ReadoutSpec
[1] https://twiki.cern.ch/twiki/bin/viewauth/CMS/UGT_readout_proposal

Comment (blwiner) [inline DataFormats/L1TGlobal/interface/GlobalAlgBlk.h]: Right, the orbitNr and bxNr were originially put in as placeholds, because
we thought these would be in the record, but they are not, so we removed
them as part of a general clean up.



On Fri, Apr 22, 2016 at 6:30 PM, tmatsush <notifications@github.com> wrote:

> In DataFormats/L1TGlobal/interface/GlobalAlgBlk.h
> <https://github.com/cms-sw/cmssw/pull/14215#discussion_r60811553>:
>
> > @@ -92,12 +90,6 @@ class GlobalAlgBlk
> >
> >  private:
> >
> > -    /// orbit number
> > -    int m_orbitNr;
> > -
> > -    /// bunch cross number of the actual bx
> > -    int m_bxNr;
> > -
>
> uGT read-out block does not [0] and will not [1] contain orbit/bx number.
> [0]
> https://twiki.cern.ch/twiki/bin/viewauth/CMS/GlobalTriggerUpgrade-ReadoutSpec
> [1] https://twiki.cern.ch/twiki/bin/viewauth/CMS/UGT_readout_proposal
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14215/files/7fc1c32245bf180c47169d08e00adfb0fdadbf08#r60811553>
>



-- 
----------------------------------------
B. Winer

**Comment by Martin-Grunewald** [comment]
This PR has merge conflicts....

**Comment by mulhearn** [comment]
Yep, working on it...

**Comment by mulhearn** [comment]
Uggh, this not working... apologies.

**Comment by mulhearn** [comment]
There's a hysteresis effect of git PRs that I don't understand.  The exact same branch with a different name, in a new PR, behaves responsibly.  We are now at cms-sw#14282.  For the rest of you, sorry for the spam.
